src/Raven.Server/RavenServer.cs

```diff
@@ -2339,41 +2340,67 @@ private async Task<bool> DispatchServerWideTcpConnection(TcpConnectionOptions tc
         {
-            var databaseLoadingTask = ServerStore.DatabasesLandlord.TryGetOrCreateResourceStore(header.DatabaseName);
-            if (databaseLoadingTask == null)
-            {
-                DatabaseDoesNotExistException.Throw(header.DatabaseName);
-                return true;
-            }
+            var result = ServerStore.DatabasesLandlord.TryGetOrCreateDatabase(header.DatabaseName);
+            switch (result.DatabaseStatus)
+            {
+                case DatabasesLandlord.DatabaseSearchResult.Status.Sharded:
+                    tcp.ShardedContext = result.ShardedContext;
+                  break;
+                case DatabasesLandlord.DatabaseSearchResult.Status.Database:
+                    var databaseLoadingTask = result.DatabaseTask;
+                    if (databaseLoadingTask.IsCompleted == false)
+                    {
+                        var resultingTask = await Task.WhenAny(databaseLoadingTask, Task.Delay(ServerStore.DatabasesLandlord.DatabaseLoadTimeout));
+                        if (resultingTask != databaseLoadingTask)
+                        {
+                            ThrowTimeoutOnDatabaseLoad(header);
+                        }
+                    }
 
-            var databaseLoadTimeout = ServerStore.DatabasesLandlord.DatabaseLoadTimeout;
+                    tcp.DocumentDatabase = await databaseLoadingTask;
+                    if (tcp.DocumentDatabase == null)
+                        DatabaseDoesNotExistException.Throw(header.DatabaseName);
 
-            if (databaseLoadingTask.IsCompleted == false)
-            {
-                var resultingTask = await Task.WhenAny(databaseLoadingTask, Task.Delay(databaseLoadTimeout));
-                if (resultingTask != databaseLoadingTask)
-                {
-                    ThrowTimeoutOnDatabaseLoad(header);
-                }
-            }
+                    Debug.Assert(tcp.DocumentDatabase != null);
 
-            tcp.DocumentDatabase = await databaseLoadingTask;
-            if (tcp.DocumentDatabase == null)
-                DatabaseDoesNotExistException.Throw(header.DatabaseName);
+                    if (tcp.DocumentDatabase.DatabaseShutdown.IsCancellationRequested)
+                        ThrowDatabaseShutdown(tcp.DocumentDatabase);
 
-            Debug.Assert(tcp.DocumentDatabase != null);
+                    tcp.DocumentDatabase.RunningTcpConnections.Add(tcp);
 
-            if (tcp.DocumentDatabase.DatabaseShutdown.IsCancellationRequested)
-                ThrowDatabaseShutdown(tcp.DocumentDatabase);
+                    if (ShardHelper.IsShardedName(tcp.DocumentDatabase.Name))
+                    {
+                        var shardedName = ShardHelper.ToDatabaseName(tcp.DocumentDatabase.Name);
+                        var shardedResult = ServerStore.DatabasesLandlord.TryGetOrCreateDatabase(shardedName);
+                        tcp.ShardedContext = shardedResult.ShardedContext;
```

Comment:
why do we need a shard context for the tcp connection on a shard instance?

Reply:
I only use it to allocate the shard data, I guess I can use the database name for that information